File: static/js/batch_monitor.js
// Batch Monitoring JavaScript

class BatchMonitor {
    constructor() {
        this.currentBatchId = null;
        this.websocket = null;
        this.reconnectAttempts = 0;
        this.maxReconnectAttempts = 5;
    }

    async checkCurrentBatch() {
        try {
            const response = await fetch('/api/batch/current');
            const data = await response.json();
            
            if (data.batch_id) {
                this.currentBatchId = data.batch_id;
                this.showMonitor(data);
                this.connectWebSocket(data.batch_id);
            } else {
                this.hideMonitor();
            }
        } catch (error) {
            console.error('Error checking current batch:', error);
        }
    }

    showMonitor(batchData) {
        const monitorDiv = document.getElementById('batch-monitor');
        if (!monitorDiv) return;

        monitorDiv.style.display = 'block';
        this.updateUI(batchData);
    }

    hideMonitor() {
        const monitorDiv = document.getElementById('batch-monitor');
        if (monitorDiv) {
            monitorDiv.style.display = 'none';
        }
    }

    updateUI(data) {
        // Update status badge
        const statusBadge = document.getElementById('batch-status');
        if (statusBadge) {
            statusBadge.textContent = data.status.toUpperCase();
            statusBadge.className = `badge badge-${this.getStatusColor(data.status)}`;
        }

        // Update progress
        const progressBar = document.getElementById('batch-progress-bar');
        const progressText = document.getElementById('batch-progress-text');
        if (progressBar && data.total > 0) {
            const percentage = Math.round((data.processed / data.total) * 100);
            progressBar.style.width = percentage + '%';
            progressText.textContent = `${data.processed} / ${data.total} (${percentage}%)`;
        }

        // Update counts
        if (document.getElementById('batch-successful')) {
            document.getElementById('batch-successful').textContent = data.successful || 0;
        }
        if (document.getElementById('batch-failed')) {
            document.getElementById('batch-failed').textContent = data.failed || 0;
        }

        // Update current call info
        if (data.current_customer_name) {
            document.getElementById('current-customer').textContent = data.current_customer_name;
            document.getElementById('current-company').textContent = data.current_company_name;
            document.getElementById('current-call-info').style.display = 'block';
        } else {
            document.getElementById('current-call-info').style.display = 'none';
        }

        // Update transcript
        if (data.live_transcript) {
            document.getElementById('live-transcript-content').textContent = data.live_transcript;
            document.getElementById('live-transcript-section').style.display = 'block';
        }

        // Update logs
        if (data.logs && data.logs.length > 0) {
            this.updateLogs(data.logs);
        }

        // Update control buttons
        this.updateControlButtons(data.status);
    }

    updateLogs(logs) {
        const logsContainer = document.getElementById('batch-logs');
        if (!logsContainer) return;

        // Show last 20 logs
        const recentLogs = logs.slice(-20).reverse();
        
        logsContainer.innerHTML = recentLogs.map(log => `
            <div class="log-entry log-${log.level}">
                <span class="log-time">${new Date(log.timestamp).toLocaleTimeString()}</span>
                <span class="log-level">[${log.level.toUpperCase()}]</span>
                <span class="log-message">${log.message}</span>
            </div>
        `).join('');
    }

    updateControlButtons(status) {
        const pauseBtn = document.getElementById('batch-pause-btn');
        const resumeBtn = document.getElementById('batch-resume-btn');
        const stopBtn = document.getElementById('batch-stop-btn');

        if (status === 'running') {
            if (pauseBtn) pauseBtn.style.display = 'inline-block';
            if (resumeBtn) resumeBtn.style.display = 'none';
            if (stopBtn) stopBtn.style.display = 'inline-block';
        } else if (status === 'paused') {
            if (pauseBtn) pauseBtn.style.display = 'none';
            if (resumeBtn) resumeBtn.style.display = 'inline-block';
            if (stopBtn) stopBtn.style.display = 'inline-block';
        } else {
            if (pauseBtn) pauseBtn.style.display = 'none';
            if (resumeBtn) resumeBtn.style.display = 'none';
            if (stopBtn) stopBtn.style.display = 'none';
        }
    }

    getStatusColor(status) {
        const colors = {
            'running': 'success',
            'paused': 'warning',
            'stopped': 'danger',
            'completed': 'info',
            'failed': 'danger',
            'idle': 'gray'
        };
        return colors[status] || 'gray';
    }

    connectWebSocket(batchId) {
        if (this.websocket) {
            this.websocket.close();
        }

        const protocol = window.location.protocol === 'https:' ? 'wss:' : 'ws:';
        const wsUrl = `${protocol}//${window.location.host}/api/batch/ws/${batchId}`;

        this.websocket = new WebSocket(wsUrl);

        this.websocket.onopen = () => {
            console.log('WebSocket connected');
            this.reconnectAttempts = 0;
        };

        this.websocket.onmessage = (event) => {
            const data = JSON.parse(event.data);
            this.handleWebSocketMessage(data);
        };

        this.websocket.onerror = (error) => {
            console.error('WebSocket error:', error);
        };

        this.websocket.onclose = () => {
            console.log('WebSocket closed');
            this.attemptReconnect(batchId);
        };

        // Keep connection alive
        this.pingInterval = setInterval(() => {
            if (this.websocket && this.websocket.readyState === WebSocket.OPEN) {
                this.websocket.send('ping');
            }
        }, 30000); // Ping every 30 seconds
    }

    handleWebSocketMessage(data) {
        switch (data.type) {
            case 'initial':
                this.updateUI(data.data);
                break;
            case 'progress':
                this.updateProgress(data);
                break;
            case 'current_call':
                this.updateCurrentCall(data);
                break;
            case 'log':
                this.addLog(data.log);
                break;
            case 'transcript':
                this.updateTranscript(data.transcript);
                break;
            case 'status':
                this.updateStatus(data.status);
                break;
        }
    }

    updateProgress(data) {
        const progressBar = document.getElementById('batch-progress-bar');
        const progressText = document.getElementById('batch-progress-text');
        if (progressBar && data.total > 0) {
            const percentage = Math.round((data.processed / data.total) * 100);
            progressBar.style.width = percentage + '%';
            progressText.textContent = `${data.processed} / ${data.total} (${percentage}%)`;
        }
        
        if (document.getElementById('batch-successful')) {
            document.getElementById('batch-successful').textContent = data.successful;
        }
        if (document.getElementById('batch-failed')) {
            document.getElementById('batch-failed').textContent = data.failed;
        }
    }

    updateCurrentCall(data) {
        document.getElementById('current-customer').textContent = data.customer_name;
        document.getElementById('current-company').textContent = data.company_name;
        document.getElementById('current-call-info').style.display = 'block';
    }

    addLog(log) {
        const logsContainer = document.getElementById('batch-logs');
        if (!logsContainer) return;

        const logEntry = document.createElement('div');
        logEntry.className = `log-entry log-${log.level}`;
        logEntry.innerHTML = `
            <span class="log-time">${new Date(log.timestamp).toLocaleTimeString()}</span>
            <span class="log-level">[${log.level.toUpperCase()}]</span>
            <span class="log-message">${log.message}</span>
        `;
        
        logsContainer.insertBefore(logEntry, logsContainer.firstChild);
        
        // Keep only last 20 logs
        while (logsContainer.children.length > 20) {
            logsContainer.removeChild(logsContainer.lastChild);
        }
    }

    updateTranscript(transcript) {
        document.getElementById('live-transcript-content').textContent = transcript;
        document.getElementById('live-transcript-section').style.display = 'block';
        
        // Auto-scroll to bottom
        const transcriptDiv = document.getElementById('live-transcript-content');
        transcriptDiv.scrollTop = transcriptDiv.scrollHeight;
    }

    updateStatus(status) {
        const statusBadge = document.getElementById('batch-status');
        if (statusBadge) {
            statusBadge.textContent = status.toUpperCase();
            statusBadge.className = `badge badge-${this.getStatusColor(status)}`;
        }
        
        this.updateControlButtons(status);
        
        // If completed or stopped, close monitor after a delay
        if (status === 'completed' || status === 'stopped') {
            setTimeout(() => {
                this.cleanup();
                this.hideMonitor();
            }, 5000);
        }
    }

    attemptReconnect(batchId) {
        if (this.reconnectAttempts < this.maxReconnectAttempts) {
            this.reconnectAttempts++;
            console.log(`Reconnecting... Attempt ${this.reconnectAttempts}`);
            setTimeout(() => {
                this.connectWebSocket(batchId);
            }, 2000 * this.reconnectAttempts);
        }
    }

    async pauseBatch() {
        if (!this.currentBatchId) return;
        
        try {
            const response = await fetch(`/api/batch/${this.currentBatchId}/pause`, {
                method: 'POST'
            });
            if (response.ok) {
                console.log('Batch paused');
            }
        } catch (error) {
            console.error('Error pausing batch:', error);
            alert('Failed to pause batch');
        }
    }

    async resumeBatch() {
        if (!this.currentBatchId) return;
        
        try {
            const response = await fetch(`/api/batch/${this.currentBatchId}/resume`, {
                method: 'POST'
            });
            if (response.ok) {
                console.log('Batch resumed');
            }
        } catch (error) {
            console.error('Error resuming batch:', error);
            alert('Failed to resume batch');
        }
    }

    async stopBatch() {
        if (!this.currentBatchId) return;
        
        if (!confirm('Are you sure you want to stop the batch? This cannot be undone.')) {
            return;
        }
        
        try {
            const response = await fetch(`/api/batch/${this.currentBatchId}/stop`, {
                method: 'POST'
            });
            if (response.ok) {
                console.log('Batch stopped');
            }
        } catch (error) {
            console.error('Error stopping batch:', error);
            alert('Failed to stop batch');
        }
    }

    cleanup() {
        if (this.websocket) {
            this.websocket.close();
            this.websocket = null;
        }
        if (this.pingInterval) {
            clearInterval(this.pingInterval);
            this.pingInterval = null;
        }
        this.currentBatchId = null;
    }
}

// Global batch monitor instance
const batchMonitor = new BatchMonitor();

// Initialize on page load
document.addEventListener('DOMContentLoaded', () => {
    if (document.getElementById('batch-monitor')) {
        batchMonitor.checkCurrentBatch();
        
        // Attach button handlers
        const pauseBtn = document.getElementById('batch-pause-btn');
        const resumeBtn = document.getElementById('batch-resume-btn');
        const stopBtn = document.getElementById('batch-stop-btn');
        
        if (pauseBtn) pauseBtn.onclick = () => batchMonitor.pauseBatch();
        if (resumeBtn) resumeBtn.onclick = () => batchMonitor.resumeBatch();
        if (stopBtn) stopBtn.onclick = () => batchMonitor.stopBatch();
    }
});

File: services/batch_monitor.py
"""
Batch monitoring service for real-time tracking and control.
"""
from datetime import datetime
from typing import Optional, Dict, List
import uuid
import logging

logger = logging.getLogger(__name__)


class BatchMonitor:
    """Singleton service to monitor and control batch processing."""
    
    def __init__(self):
        self.current_batch_id: Optional[str] = None
        self.should_pause = False
        self.should_stop = False
        self._listeners = []  # WebSocket connections
    
    def create_batch(self, total_count: int, triggered_by: str = None) -> str:
        """Create a new batch process."""
        from database import get_db_context
        from models import BatchProcess, BatchStatus
        
        batch_id = f"batch_{uuid.uuid4().hex[:12]}"
        
        with get_db_context() as db:
            batch = BatchProcess(
                batch_id=batch_id,
                status=BatchStatus.RUNNING,
                total_verifications=total_count,
                started_at=datetime.utcnow(),
                triggered_by=triggered_by,
                logs=[]
            )
            db.add(batch)
            db.commit()
        
        self.current_batch_id = batch_id
        self.should_pause = False
        self.should_stop = False
        
        logger.info(f"Created batch {batch_id} with {total_count} verifications")
        return batch_id
    
    def update_progress(self, batch_id: str, processed: int, successful: int, failed: int):
        """Update batch progress."""
        from database import get_db_context
        from models import BatchProcess
        
        with get_db_context() as db:
            batch = db.query(BatchProcess).filter(BatchProcess.batch_id == batch_id).first()
            if batch:
                batch.processed_count = processed
                batch.successful_count = successful
                batch.failed_count = failed
                batch.updated_at = datetime.utcnow()
                db.commit()
                
                # Notify listeners
                self._notify_listeners({
                    "type": "progress",
                    "batch_id": batch_id,
                    "processed": processed,
                    "successful": successful,
                    "failed": failed,
                    "total": batch.total_verifications
                })
    
    def set_current_call(self, batch_id: str, verification_id: str, call_sid: str, 
                        customer_name: str, company_name: str):
        """Update current call information."""
        from database import get_db_context
        from models import BatchProcess
        
        with get_db_context() as db:
            batch = db.query(BatchProcess).filter(BatchProcess.batch_id == batch_id).first()
            if batch:
                batch.current_verification_id = verification_id
                batch.current_call_sid = call_sid
                batch.current_customer_name = customer_name
                batch.current_company_name = company_name
                db.commit()
                
                # Notify listeners
                self._notify_listeners({
                    "type": "current_call",
                    "batch_id": batch_id,
                    "verification_id": verification_id,
                    "call_sid": call_sid,
                    "customer_name": customer_name,
                    "company_name": company_name
                })
    
    def add_log(self, batch_id: str, level: str, message: str):
        """Add a log entry to the batch."""
        from database import get_db_context
        from models import BatchProcess
        
        log_entry = {
            "timestamp": datetime.utcnow().isoformat(),
            "level": level,
            "message": message
        }
        
        with get_db_context() as db:
            batch = db.query(BatchProcess).filter(BatchProcess.batch_id == batch_id).first()
            if batch:
                logs = batch.logs or []
                logs.append(log_entry)
                batch.logs = logs
                db.commit()
                
                # Notify listeners
                self._notify_listeners({
                    "type": "log",
                    "batch_id": batch_id,
                    "log": log_entry
                })
    
    def update_transcript(self, batch_id: str, transcript: str):
        """Update live transcript."""
        from database import get_db_context
        from models import BatchProcess
        
        with get_db_context() as db:
            batch = db.query(BatchProcess).filter(BatchProcess.batch_id == batch_id).first()
            if batch:
                batch.live_transcript = transcript
                db.commit()
                
                # Notify listeners
                self._notify_listeners({
                    "type": "transcript",
                    "batch_id": batch_id,
                    "transcript": transcript
                })
    
    def pause_batch(self, batch_id: str):
        """Pause the batch."""
        from database import get_db_context
        from models import BatchProcess, BatchStatus
        
        with get_db_context() as db:
            batch = db.query(BatchProcess).filter(BatchProcess.batch_id == batch_id).first()
            if batch:
                batch.status = BatchStatus.PAUSED
                batch.paused_at = datetime.utcnow()
                db.commit()
        
        self.should_pause = True
        logger.info(f"Paused batch {batch_id}")
        
        self._notify_listeners({
            "type": "status",
            "batch_id": batch_id,
            "status": "paused"
        })
    
    def resume_batch(self, batch_id: str):
        """Resume the batch."""
        from database import get_db_context
        from models import BatchProcess, BatchStatus
        
        with get_db_context() as db:
            batch = db.query(BatchProcess).filter(BatchProcess.batch_id == batch_id).first()
            if batch:
                batch.status = BatchStatus.RUNNING
                batch.paused_at = None
                db.commit()
        
        self.should_pause = False
        logger.info(f"Resumed batch {batch_id}")
        
        self._notify_listeners({
            "type": "status",
            "batch_id": batch_id,
            "status": "running"
        })
    
    def stop_batch(self, batch_id: str):
        """Stop the batch."""
        from database import get_db_context
        from models import BatchProcess, BatchStatus
        
        with get_db_context() as db:
            batch = db.query(BatchProcess).filter(BatchProcess.batch_id == batch_id).first()
            if batch:
                batch.status = BatchStatus.STOPPED
                batch.completed_at = datetime.utcnow()
                db.commit()
        
        self.should_stop = True
        logger.info(f"Stopped batch {batch_id}")
        
        self._notify_listeners({
            "type": "status",
            "batch_id": batch_id,
            "status": "stopped"
        })
    
    def complete_batch(self, batch_id: str):
        """Mark batch as completed."""
        from database import get_db_context
        from models import BatchProcess, BatchStatus
        
        with get_db_context() as db:
            batch = db.query(BatchProcess).filter(BatchProcess.batch_id == batch_id).first()
            if batch:
                batch.status = BatchStatus.COMPLETED
                batch.completed_at = datetime.utcnow()
                db.commit()
        
        self.current_batch_id = None
        logger.info(f"Completed batch {batch_id}")
        
        self._notify_listeners({
            "type": "status",
            "batch_id": batch_id,
            "status": "completed"
        })
    
    def get_status(self, batch_id: str) -> Optional[Dict]:
        """Get current batch status."""
        from database import get_db_context
        from models import BatchProcess
        
        with get_db_context() as db:
            batch = db.query(BatchProcess).filter(BatchProcess.batch_id == batch_id).first()
            if batch:
                return {
                    "batch_id": batch.batch_id,
                    "status": batch.status.value,
                    "total": batch.total_verifications,
                    "processed": batch.processed_count,
                    "successful": batch.successful_count,
                    "failed": batch.failed_count,
                    "current_verification_id": batch.current_verification_id,
                    "current_call_sid": batch.current_call_sid,
                    "current_customer_name": batch.current_customer_name,
                    "current_company_name": batch.current_company_name,
                    "live_transcript": batch.live_transcript,
                    "logs": batch.logs or [],
                    "started_at": batch.started_at.isoformat() if batch.started_at else None,
                    "completed_at": batch.completed_at.isoformat() if batch.completed_at else None
                }
        return None
    
    def add_listener(self, callback):
        """Add a WebSocket listener."""
        self._listeners.append(callback)
    
    def remove_listener(self, callback):
        """Remove a WebSocket listener."""
        if callback in self._listeners:
            self._listeners.remove(callback)
    
    def _notify_listeners(self, data: Dict):
        """Notify all connected WebSocket listeners."""
        for listener in self._listeners:
            try:
                listener(data)
            except Exception as e:
                logger.error(f"Error notifying listener: {e}")


# Global batch monitor instance
batch_monitor = BatchMonitor()
